File: SDK/SDK/SDK_python/CodeCraft-2021/src/CodeCraft-2021.py
import numpy as np
import  sys
def main():
    # to read standard input

    CPU = 0
    MEM = 1
    NODE = 2
    IDVM = 3
    VSIZE = 4
    ADD_INDEX = 5
    NODE_A = 0
    NODE_B = 1
    VMSET = 2
    PM_INDEX = 0
    VM_NODE = 3

    def vsizeAlg(opAdd):
        opAdd[:,4] = opAdd[:,0]/np.sum(opAdd[:,0]) + opAdd[:,1]/np.sum(opAdd[:,1])

    def physizeAlg(cpu,mem):
        return cpu*mem


    targetPM = None
    def buyPM(matPM): # 记录一下每次买的型号
        return [[matPM[targetPM][CPU]/2,matPM[targetPM][MEM]/2],[matPM[targetPM][CPU]/2,matPM[targetPM][MEM]/2],set()]


    # 读取物理机和虚拟机型号
    typePM = int(input())
    matPM = dict()
    currMax = None
    for i in range(typePM):
        currLine = input().strip("()\n").split(',')
        cpu, mem, price, maintain = int(currLine[1]), int(currLine[2]), int(currLine[3]), int(currLine[4])
        matPM[currLine[0]] = [cpu, mem, price, maintain]
        if currMax == None:
            currMax = [currLine[0],physizeAlg(cpu,mem)]
        elif currMax[1] < physizeAlg(cpu,mem):
            currMax = [currLine[0],physizeAlg(cpu,mem)]
    targetPM = currMax[0]

    typeVM = int(input())
    matVM = dict()
    for i in range(typeVM):
        currLine = input().strip("()\n").split(',')
        cpu, mem, node = int(currLine[1]), int(currLine[2]), int(currLine[3])
        matVM[currLine[0]] = [cpu, mem, node]

        
    # 创建第一个物理机
    vm2pm = dict()
    pm2vm = []
    numPM = len(pm2vm)
    pm2vm.append(buyPM(matPM))

    # 读取每天的信息并处理
    days = int(input())
    for i in range(days):
        todayReq = int(input())
        opAdd = np.zeros((todayReq,6)) # cpu mem node vmID vsize index
        opAddIndex = 0
        opDel = np.zeros((todayReq,1))
        opDelIndex = 0
        # 读取本日请求
        for request in range(todayReq):
            currLine = input().strip("()\n").split(',')
            op = currLine[0]
            if op == 'add':
                buyType, vmID = currLine[1].strip(), int(currLine[2])
                opAdd[opAddIndex] = matVM[buyType] + [vmID,0,opAddIndex] # cpu mem node vmID vsize index
                opAddIndex += 1

            else:
                opDel[opDelIndex] = int(currLine[1])
                opDelIndex += 1
        
        # 删除
        for eachDel in range(opDelIndex):
            vmID = opDel[eachDel,0]
            pmvmMsg = vm2pm.pop(vmID)
            pmIndex = pmvmMsg.pop(0)
            if len(pmvmMsg) == 2:
                pm2vm[pmIndex][NODE_A][CPU] += pmvmMsg[CPU]
                pm2vm[pmIndex][NODE_A][MEM] += pmvmMsg[MEM]
                pm2vm[pmIndex][NODE_B][CPU] += pmvmMsg[CPU]
                pm2vm[pmIndex][NODE_B][MEM] += pmvmMsg[MEM]
            else:
                pm2vm[pmIndex][pmvmMsg[NODE]][CPU] += pmvmMsg[CPU]
                pm2vm[pmIndex][pmvmMsg[NODE]][MEM] += pmvmMsg[MEM]
            pm2vm[pmIndex][VMSET].remove(vmID)
                




        # 部署
        vsizeAlg(opAdd) # 计算需求的尺寸
        opAdd = opAdd[np.argsort(-opAdd[:, VSIZE])] # 将需求按尺寸降序排列
        for i in range(opAddIndex):
            for j,eachPM in enumerate(pm2vm) :
                if opAdd[i,NODE] == 0: # single node
                    whichNode = None
                    if physizeAlg(eachPM[NODE_A][CPU],eachPM[NODE_A][MEM]) <= physizeAlg(eachPM[NODE_B][CPU],eachPM[NODE_B][MEM]):
                        whichNode = NODE_B
                    else:
                        whichNode = NODE_A
                    if eachPM[whichNode][CPU] >= opAdd[i,CPU] and eachPM[whichNode][MEM] >= opAdd[i,MEM]:
                        eachPM[whichNode][CPU] -= opAdd[i,CPU]
                        eachPM[whichNode][MEM] -= opAdd[i,MEM]
                        eachPM[VMSET].add(opAdd[i,IDVM])
                        vm2pm[opAdd[i,IDVM]] = [j,opAdd[i,CPU],opAdd[i,MEM],whichNode]

                        break
                    else:
                        pass # 看下一个物理机
                else:
                    if eachPM[NODE_A][CPU] >= opAdd[i,CPU]/2 and eachPM[NODE_A][MEM] >= opAdd[i,MEM]/2 and \
                        eachPM[NODE_B][CPU] >= opAdd[i,CPU]/2 and eachPM[NODE_B][MEM] >= opAdd[i,MEM]/2:
                        eachPM[NODE_A][CPU] -= opAdd[i,CPU]/2
                        eachPM[NODE_A][MEM] -= opAdd[i,MEM]/2
                        eachPM[NODE_B][CPU] -= opAdd[i,CPU]/2
                        eachPM[NODE_B][MEM] -= opAdd[i,MEM]/2
                        eachPM[VMSET].add(opAdd[i,IDVM])
                        vm2pm[opAdd[i,IDVM]] = [j,opAdd[i,CPU]/2,opAdd[i,MEM]/2]
                        break
                    else:
                        pass # 看下一个物理机
            else:
                pm2vm.append(buyPM(matPM)) # TODO 判断是否一定放得下
                j = len(pm2vm) - 1
                if opAdd[i,NODE] == 0: # single node
                    pm2vm[j][NODE_A][CPU] -= opAdd[i,CPU]
                    pm2vm[j][NODE_A][MEM] -= opAdd[i,MEM]
                    pm2vm[j][VMSET].add(opAdd[i,IDVM])
                    vm2pm[opAdd[i,IDVM]] = [j,opAdd[i,CPU], opAdd[i,MEM], NODE_A]
                else:
                    pm2vm[j][NODE_A][CPU] -= opAdd[i,CPU]/2
                    pm2vm[j][NODE_A][MEM] -= opAdd[i,MEM]/2
                    pm2vm[j][NODE_B][CPU] -= opAdd[i,CPU]/2
                    pm2vm[j][NODE_B][MEM] -= opAdd[i,MEM]/2
                    pm2vm[j][VMSET].add(opAdd[i,IDVM])
                    vm2pm[opAdd[i,IDVM]] = [j,opAdd[i,CPU]/2,opAdd[i,MEM]/2]
        
        '''
        (purchase, 2) 本日物理机购买数目
        (NV603, 1)
        (NV604, 1)
        (migration, 0) 迁移数目
        (0, A) 部署
        (0, B)
        '''

        outPMNum = '(purchase, %d)'
        outPMType = '(%s, %d)'
        outMigration = '(migration, %d)'
        outDeploySingle = '(%d, %s)'
        outDeployDouble = '(%d)'
        print(outPMNum % (len(pm2vm) - numPM))
        numPM = len(pm2vm)

        print(outPMType % (targetPM,len(pm2vm) - numPM))

        print(outMigration % 0)

        opAdd = opAdd[np.argsort(opAdd[:, ADD_INDEX])]
        for i in range(opAddIndex):
            if opAdd[i,IDVM] == 0.0: # 排除掉初始化项
                continue
            if len(vm2pm[opAdd[i,IDVM]]) == 3:
                print(outDeployDouble % vm2pm[opAdd[i,IDVM]][PM_INDEX])
            else:
                if vm2pm[opAdd[i,IDVM]][VM_NODE] == NODE_A: # 
                    print(outDeploySingle % (vm2pm[opAdd[i,IDVM]][PM_INDEX],'A'))
                else:
                    print(outDeploySingle % (vm2pm[opAdd[i,IDVM]][PM_INDEX],'B'))


    # to write standard output
    sys.stdout.flush()


if __name__ == "__main__":
    main()

File: CodeCraft-2021.cpp
#include "iostream"
#include "cstdlib"
#include "cstdio"
#include <unordered_map>
#include <string>
#include <sstream>
#include <array>
#include <set>
#include <vector>

using namespace std;
//* 宏定义
#define CPU			0
#define MEM			1
#define NODE 		2
#define IDVM 		3
#define VSIZE 		4
#define ADD_INDEX 	5
#define NODE_A 		0
#define NODE_B 		1
#define NODE_DOUBLE	2
#define VMSET 		2
#define VM2PM_INDEX 	3	
#define ENLARGE		100000


string targetPM;

typedef array<int, 5> arrayPM; // cpu mem price mantain vsize
int physizeAlg(arrayPM* arr) { // 计算物理机的vsize，也用于计算node的vsize
	return arr->at(CPU) + arr->at(MEM); 
}


typedef array<int, 3> arrayVM;

typedef array<int, 4> arrayVM2PM;


typedef array<int, 2> arrayNodeResource;
typedef struct PMResourceMsg_t
{
	arrayNodeResource nodeA;
	arrayNodeResource nodeB;
	set<int> vmset;
}PMResourceMsg;


PMResourceMsg buyPM(string PMName, unordered_map<string, arrayPM>* matPM) {
	PMResourceMsg msg;
	arrayNodeResource rsc = { matPM->at(PMName)[CPU] / 2,matPM->at(PMName)[MEM] / 2 };
	msg.nodeA = rsc;
	msg.nodeB = rsc;
	return msg;
}


typedef array<int, 6> arrayOpAdd;


void vsizeAlg(vector<arrayOpAdd> *opAdd){
	int sumCPU = 0,sumMEM = 0;
	for (int i=0;i<opAdd->size();++i){
		sumCPU += opAdd->at(i)[CPU];
		sumMEM += opAdd->at(i)[MEM];
	}
	for (int i=0;i<opAdd->size();++i){
		opAdd->at(i)[VSIZE] = ENLARGE*(opAdd->at(i)[CPU]/sumCPU + opAdd->at(i)[MEM]/sumMEM);
	}
}

void quicksortUP(vector<arrayOpAdd> *opAdd,int key,int begin,int end){
	int mid = begin;
	int begin_ = begin,end_ = end;
	begin += 1;
	while (begin <  end){
		while(begin<end){
			if (opAdd->at(end)[VSIZE] >= opAdd->at(mid)[VSIZE]){
				end -= 1;
			}
			else{
				break;
			}
		}
		while(begin<end){
			if (opAdd->at(begin)[VSIZE] <= opAdd->at(mid)[VSIZE]){
				begin += 1;
			}
			else{
				break;
			}
		}
		auto temp = opAdd->at(begin);
		opAdd->at(begin) = opAdd->at(end);
		opAdd->at(end) = temp;

	}
	auto temp = opAdd->at(begin);
	opAdd->at(begin) = opAdd->at(mid);
	opAdd->at(mid) = temp;

	quicksortUP(opAdd,key,begin_,mid-1);
	quicksortUP(opAdd,key,mid+1,end_);


}

void quicksortDOWN(vector<arrayOpAdd> *opAdd,int key,int begin,int end){
	int mid = begin;
	int begin_ = begin,end_ = end;
	begin += 1;
	while (begin <  end){
		while(begin<end){
			if (opAdd->at(end)[VSIZE] <= opAdd->at(mid)[VSIZE]){
				end -= 1;
			}
			else{
				break;
			}
		}
		while(begin<end){
			if (opAdd->at(begin)[VSIZE] >= opAdd->at(mid)[VSIZE]){
				begin += 1;
			}
			else{
				break;
			}
		}
		auto temp = opAdd->at(begin);
		opAdd->at(begin) = opAdd->at(end);
		opAdd->at(end) = temp;

	}
	auto temp = opAdd->at(begin);
	opAdd->at(begin) = opAdd->at(mid);
	opAdd->at(mid) = temp;

	quicksortUP(opAdd,key,begin_,mid-1);
	quicksortUP(opAdd,key,mid+1,end_);


}

bool canSetSingle(arrayNodeResource* nodeMSG,arrayOpAdd* opaddMSG){
	if (nodeMSG->at(CPU)>= opaddMSG->at(CPU) && nodeMSG->at(MEM)>= opaddMSG->at(MEM)){
		return true;
	}
	else{
		return false;
	}

}

bool canSetDouble(arrayNodeResource* nodeMSG,arrayOpAdd* opaddMSG){
	if (nodeMSG->at(CPU)>= opaddMSG->at(CPU)/2 && nodeMSG->at(MEM)>= opaddMSG->at(MEM)/2){
		return true;
	}
	else{
		return false;
	}

}
int main()
{
	// TODO:read standard input
	// TODO:process

	int typePM;
	char _n;// 读取末尾换行符

	cin >> typePM;
	unordered_map<string, arrayPM> matPM;
	int currMaxVsize = -1;
	string currMaxName;
	//* 读取物理机数据并选出 vsize 最大的物理机
	for (int i = 0; i < typePM; ++i) {
		cin >> _n; // 读取 (
		string currLine;
		getline(cin, currLine);
		istringstream is(currLine);
		string pmName;
		is >> pmName;
		pmName.erase(pmName.end() - 1);
		int cpu, mem, price, maintain;

		is >> cpu >> _n >> mem >> _n >> price >> _n >> maintain >> _n;
		arrayPM pmVal = { cpu, mem, price, maintain,0 };
		pmVal[VSIZE] = physizeAlg(&pmVal);
		matPM[pmName] = pmVal;

		if (currMaxVsize == -1) {
			currMaxVsize = matPM[pmName][VSIZE];
			currMaxName = pmName;
		}
		else if (currMaxVsize < matPM[pmName][VSIZE]) {
			currMaxVsize = matPM[pmName][VSIZE];
			currMaxName = pmName;
		}

	}

	targetPM = currMaxName;


	//* 读取虚拟机信息
	int typeVM;
	cin >> typeVM;
	unordered_map<string, arrayVM> matVM;
	for (int i = 0; i < typeVM; ++i) {
		cin >> _n; // 读取 (
		string currLine;
		getline(cin, currLine);
		istringstream is(currLine);
		string vmName;
		is >> vmName;
		vmName.erase(vmName.end() - 1);
		int cpu, mem, node;

		is >> cpu >> _n >> mem >> _n >> node >> _n;
		arrayVM vmVal = { cpu, mem, node };
		matVM[vmName] = vmVal;

	}



	unordered_map<int, arrayVM2PM> vm2pm;
	vector<PMResourceMsg> pm2vm;
	vector<arrayOpAdd> opAdd; // cpu mem node vmID vsize index
	vector<int> opDel; // vmID
	int numPM = 0;

	int days;
	cin >> days;
	for (int i = 0; i < days; ++i) {
		int todayReq;
		cin >> todayReq;
		int opAddIndex = 0, opDelIndex = 0;
		for (int request = 0; request < todayReq; ++request) {
			cin >> _n; // 读取 (
			string currLine;
			getline(cin, currLine);
			istringstream is(currLine);
			string op;
			is >> op;
			op.erase(op.end() - 1);

			int vmID;
			if (op == "add") {
				string vmType;
				is >> vmType;
				vmType.erase(vmType.end() - 1);
				is >> vmID >> _n;
				arrayOpAdd vmMsg = { matVM[vmType][CPU],matVM[vmType][MEM],matVM[vmType][NODE],vmID,0,opAddIndex };
				opAdd.push_back(vmMsg);
				opAddIndex += 1;

			}
			else {
				is >> vmID >> _n >> _n;
				opDel.push_back(vmID);
				opDelIndex += 1;

			}
		}

		//* 删除
		for (int i = 0;i<opDelIndex;++i){
			int vmID = opDel[i];
			arrayVM2PM pmvmMsg = vm2pm[vmID];
			vm2pm.erase(vmID);
			int pmIndex = pmvmMsg[VM2PM_INDEX];
			if (pmvmMsg[NODE] == NODE_DOUBLE){
				pm2vm[pmIndex].nodeA[CPU] += pmvmMsg[CPU];
				pm2vm[pmIndex].nodeA[MEM] += pmvmMsg[MEM];
				pm2vm[pmIndex].nodeB[CPU] += pmvmMsg[CPU];
				pm2vm[pmIndex].nodeB[MEM] += pmvmMsg[MEM];
			}
			else{
				if (pmvmMsg[NODE] == NODE_A){
					pm2vm[pmIndex].nodeA[CPU] += pmvmMsg[CPU];
					pm2vm[pmIndex].nodeA[MEM] += pmvmMsg[MEM];
				}
				else{
					pm2vm[pmIndex].nodeB[CPU] += pmvmMsg[CPU];
					pm2vm[pmIndex].nodeB[MEM] += pmvmMsg[MEM];
				}
			}
			pm2vm[pmIndex].vmset.erase(vmID);
		}

		//* 迁移


		//* 部署
		// 计算需求的 vsize
		vsizeAlg(&opAdd);
		quicksortUP(&opAdd,VSIZE,0,opAdd.size());
		for (int i=0;i<opAddIndex;++i){
			bool isSet = false;
			for (int j=0;j<pm2vm.size();++j){ // 找到一个可以放下请求的物理机
				if (opAdd[i][NODE] == 0){ // single node
					int whichNode = 0;
					if (physizeAlg((arrayPM*)&(pm2vm[j].nodeA)) <= physizeAlg((arrayPM*)&(pm2vm[j].nodeB))){
						whichNode = NODE_B;
					}
					else{
						whichNode = NODE_A;
					}

					if (whichNode == NODE_A && canSetSingle(&(pm2vm[j].nodeA),&(opAdd[i]))){
						pm2vm[j].nodeA[CPU] -= opAdd[i][CPU];
						pm2vm[j].nodeA[MEM] -= opAdd[i][MEM];
						pm2vm[j].vmset.insert(opAdd[i][IDVM]);
						arrayVM2PM vmMsg = {opAdd[i][CPU],opAdd[i][MEM],whichNode,j};
						vm2pm[opAdd[i][IDVM]] = vmMsg;
						isSet = true;
					}
					else if (whichNode == NODE_B && canSetSingle(&(pm2vm[j].nodeB),&(opAdd[i]))){
						pm2vm[j].nodeB[CPU] -= opAdd[i][CPU];
						pm2vm[j].nodeB[MEM] -= opAdd[i][MEM];
						pm2vm[j].vmset.insert(opAdd[i][IDVM]);
						arrayVM2PM vmMsg = {opAdd[i][CPU],opAdd[i][MEM],whichNode,j};
						vm2pm[opAdd[i][IDVM]] = vmMsg;
						isSet = true;
					}
				}
				else{ // 双节点部署
					if (canSetDouble(&(pm2vm[j].nodeA),&(opAdd[i])) && canSetDouble(&(pm2vm[j].nodeB),&(opAdd[i]))){
						pm2vm[j].nodeA[CPU] -= opAdd[i][CPU]/2;
						pm2vm[j].nodeA[MEM] -= opAdd[i][MEM]/2;
						pm2vm[j].nodeB[CPU] -= opAdd[i][CPU]/2;
						pm2vm[j].nodeB[MEM] -= opAdd[i][MEM]/2;
						pm2vm[j].vmset.insert(opAdd[i][IDVM]);
						arrayVM2PM vmMsg = {opAdd[i][CPU],opAdd[i][MEM],NODE_DOUBLE,j};
						vm2pm[opAdd[i][IDVM]] = vmMsg;
						isSet = true;

					}
				}
			}
			if (isSet == false){ // 买一台物理机
				pm2vm.push_back(buyPM(targetPM, &matPM)); // TODO 判断是否一定放得下
				int j = pm2vm.size() - 1;
				if (opAdd[i][NODE] == 0){ // single node
					int whichNode = NODE_A;
					pm2vm[j].nodeA[CPU] -= opAdd[i][CPU];
					pm2vm[j].nodeA[MEM] -= opAdd[i][MEM];
					pm2vm[j].vmset.insert(opAdd[i][IDVM]);
					arrayVM2PM vmMsg = {opAdd[i][CPU],opAdd[i][MEM],whichNode,j};
					vm2pm[opAdd[i][IDVM]] = vmMsg;

				}
				else{ // 双节点部署
					pm2vm[j].nodeA[CPU] -= opAdd[i][CPU]/2;
					pm2vm[j].nodeA[MEM] -= opAdd[i][MEM]/2;
					pm2vm[j].nodeB[CPU] -= opAdd[i][CPU]/2;
					pm2vm[j].nodeB[MEM] -= opAdd[i][MEM]/2;
					pm2vm[j].vmset.insert(opAdd[i][IDVM]);
					arrayVM2PM vmMsg = {opAdd[i][CPU],opAdd[i][MEM],NODE_DOUBLE,j};
					vm2pm[opAdd[i][IDVM]] = vmMsg;
				}
			}
		}


		char* outPMTypeAll = "(purchase, %d)";
		char* outPMType = "(%s, %d)";
		char* outMigration = "(migration, %d)";
		char* outDeploySingle = "(%d, %s)";
		char* outDeployDouble = "(%d)";
		int needtoBuy = pm2vm.size() - numPM;
		if (needtoBuy == 0){
			printf(outPMTypeAll , 0);
		}
		else{
			printf(outPMTypeAll , 1);
			printf(outPMType , targetPM.data(),needtoBuy);
		}
		numPM = pm2vm.size();

		printf(outMigration , 0);

		quicksortDOWN(&opAdd,VM2PM_INDEX,0,opAdd.size());
		for (int i=0;i<todayReq;++i){
			if (opAdd[i][IDVM] != 0){
				int vmID = opAdd[i][IDVM];
				arrayVM2PM pmvmMsg = vm2pm[vmID];
				if (pmvmMsg[NODE] == NODE_DOUBLE){
					printf(outDeployDouble , vm2pm[vmID][VM2PM_INDEX]);
				}
				else{
					if (pmvmMsg[NODE] == NODE_A){
						printf(outDeploySingle , vm2pm[vmID][VM2PM_INDEX],'A');
					}
					else{
						printf(outDeploySingle , vm2pm[vmID][VM2PM_INDEX],'B');
					}
				}
			}
		}

	}


	// TODO:write standard output
	// TODO:fflush(stdout);

	return 0;
}


